File: src/morphic/inspector/integrity.py
"""
File integrity checking for images and videos.

Uses Pillow's ``verify()`` / ``load()`` for images and ``ffprobe``
for videos.
"""

from __future__ import annotations

import logging
import os
import subprocess
from concurrent.futures import ThreadPoolExecutor, as_completed

from PIL import Image

from morphic.shared.constants import IMAGE_EXTENSIONS, VIDEO_EXTENSIONS
from morphic.shared.utils import (
    find_files_by_extension,
    format_file_size,
    is_image,
    is_video,
)

logger = logging.getLogger(__name__)


def check_image(path: str) -> dict:
    """Validate an image file's integrity.

    Parameters
    ----------
    path : str
        Path to the image file.

    Returns
    -------
    dict
        ``{"path", "valid", "error", "size", "size_formatted",
        "width", "height", "format"}``
    """
    result: dict = {
        "path": path,
        "valid": False,
        "error": None,
        "size": 0,
        "size_formatted": "0 B",
        "width": 0,
        "height": 0,
        "format": None,
        "type": "image",
    }

    if not os.path.isfile(path):
        result["error"] = "File not found"
        return result

    result["size"] = os.path.getsize(path)
    result["size_formatted"] = format_file_size(result["size"])

    if result["size"] == 0:
        result["error"] = "Zero-byte file"
        return result

    try:
        # First pass: verify structure
        img = Image.open(path)
        result["format"] = img.format
        result["width"] = img.width
        result["height"] = img.height
        img.verify()

        # Second pass: actually decode all pixels
        img = Image.open(path)
        img.load()

        result["valid"] = True
    except Exception as e:
        result["error"] = str(e)

    return result


def check_video(path: str) -> dict:
    """Validate a video file using ffprobe.

    Parameters
    ----------
    path : str
        Path to the video file.

    Returns
    -------
    dict
        ``{"path", "valid", "error", "size", "size_formatted",
        "width", "height", "duration", "codec"}``
    """
    result: dict = {
        "path": path,
        "valid": False,
        "error": None,
        "size": 0,
        "size_formatted": "0 B",
        "width": 0,
        "height": 0,
        "duration": 0.0,
        "codec": None,
        "type": "video",
    }

    if not os.path.isfile(path):
        result["error"] = "File not found"
        return result

    result["size"] = os.path.getsize(path)
    result["size_formatted"] = format_file_size(result["size"])

    if result["size"] == 0:
        result["error"] = "Zero-byte file"
        return result

    try:
        cmd = [
            "ffprobe",
            "-v", "error",
            "-select_streams", "v:0",
            "-show_entries",
            "stream=codec_name,width,height,duration",
            "-of", "csv=p=0",
            path,
        ]
        proc = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=30,
        )

        if proc.returncode != 0:
            stderr = proc.stderr.strip()
            result["error"] = stderr or f"ffprobe exit code {proc.returncode}"
            return result

        output = proc.stdout.strip()
        if not output:
            result["error"] = "No video stream found"
            return result

        parts = output.split(",")
        if len(parts) >= 1:
            result["codec"] = parts[0]
        if len(parts) >= 2:
            result["width"] = int(parts[1]) if parts[1] else 0
        if len(parts) >= 3:
            result["height"] = int(parts[2]) if parts[2] else 0
        if len(parts) >= 4 and parts[3]:
            try:
                result["duration"] = float(parts[3])
            except ValueError:
                pass

        result["valid"] = True

    except FileNotFoundError:
        result["error"] = "ffprobe not found (install ffmpeg)"
    except subprocess.TimeoutExpired:
        result["error"] = "ffprobe timed out"
    except Exception as e:
        result["error"] = str(e)

    return result


def check_files(
    folder: str,
    max_workers: int = 4,
) -> list[dict]:
    """Check integrity of all media files in a folder.

    Parameters
    ----------
    folder : str
        Root folder to scan.
    max_workers : int
        Number of threads for parallel checking.

    Returns
    -------
    list[dict]
        Per-file integrity results.
    """
    all_ext = IMAGE_EXTENSIONS | VIDEO_EXTENSIONS
    paths = find_files_by_extension(folder, all_ext)
    results: list[dict] = []

    with ThreadPoolExecutor(max_workers=max_workers) as pool:
        futures = {}
        for path in paths:
            if is_image(path):
                futures[pool.submit(check_image, path)] = path
            elif is_video(path):
                futures[pool.submit(check_video, path)] = path

        for future in as_completed(futures):
            try:
                results.append(future.result())
            except Exception as e:
                results.append({
                    "path": futures[future],
                    "valid": False,
                    "error": str(e),
                })

    return sorted(results, key=lambda r: r["path"])

File: internal/inspector/integrity.go
package inspector

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"sync"

	"github.com/disintegration/imaging"
	"github.com/exterex/morphic/internal/shared"
)

// IntegrityResult holds the outcome of a single file check.
type IntegrityResult struct {
	Path          string  `json:"path"`
	Valid         bool    `json:"valid"`
	Error         string  `json:"error,omitempty"`
	Size          int64   `json:"size"`
	SizeFormatted string  `json:"size_formatted"`
	Width         int     `json:"width"`
	Height        int     `json:"height"`
	Format        string  `json:"format,omitempty"`
	Duration      float64 `json:"duration,omitempty"`
	Codec         string  `json:"codec,omitempty"`
	Type          string  `json:"type"`
	Filename      string  `json:"filename,omitempty"`
	Directory     string  `json:"directory,omitempty"`
}

// CheckImage validates an image file by opening and decoding it.
func CheckImage(path string) IntegrityResult {
	r := IntegrityResult{Path: path, Type: "image"}

	info, err := os.Stat(path)
	if err != nil {
		r.Error = "File not found"
		return r
	}
	r.Size = info.Size()
	r.SizeFormatted = shared.FormatFileSize(r.Size)

	if r.Size == 0 {
		r.Error = "Zero-byte file"
		return r
	}

	ext := strings.ToLower(filepath.Ext(path))
	if alias, ok := shared.Aliases[ext]; ok {
		ext = alias
	}

	if ext == ".avif" {
		// Use ffprobe for AVIF validation because imaging may not decode it.
		cmd := exec.Command("ffprobe", "-v", "error", "-select_streams", "v:0", "-show_entries", "stream=codec_name,width,height", "-of", "csv=p=0", path)
		out, err := cmd.Output()
		if err != nil {
			r.Error = err.Error()
			return r
		}
		parts := strings.Split(strings.TrimSpace(string(out)), ",")
		if len(parts) >= 2 {
			r.Width, _ = strconv.Atoi(parts[1])
		}
		if len(parts) >= 3 {
			r.Height, _ = strconv.Atoi(parts[2])
		}
		if len(parts) >= 1 {
			r.Format = parts[0]
		}
		r.Valid = true
		return r
	}

	img, err := imaging.Open(path)
	if err != nil {
		r.Error = err.Error()
		return r
	}

	bounds := img.Bounds()
	r.Width = bounds.Dx()
	r.Height = bounds.Dy()
	r.Valid = true
	return r
}

// CheckVideo validates a video file using ffprobe.
func CheckVideo(path string) IntegrityResult {
	r := IntegrityResult{Path: path, Type: "video"}

	info, err := os.Stat(path)
	if err != nil {
		r.Error = "File not found"
		return r
	}
	r.Size = info.Size()
	r.SizeFormatted = shared.FormatFileSize(r.Size)

	if r.Size == 0 {
		r.Error = "Zero-byte file"
		return r
	}

	cmd := exec.Command("ffprobe",
		"-v", "error",
		"-select_streams", "v:0",
		"-show_entries", "stream=codec_name,width,height,duration",
		"-of", "csv=p=0",
		path,
	)

	out, err := cmd.Output()
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			stderr := strings.TrimSpace(string(exitErr.Stderr))
			if stderr != "" {
				r.Error = stderr
			} else {
				r.Error = fmt.Sprintf("ffprobe exit code %d", exitErr.ExitCode())
			}
		} else {
			r.Error = err.Error()
		}
		return r
	}

	output := strings.TrimSpace(string(out))
	if output == "" {
		r.Error = "No video stream found"
		return r
	}

	parts := strings.Split(output, ",")
	if len(parts) >= 1 {
		r.Codec = parts[0]
	}
	if len(parts) >= 2 {
		r.Width, _ = strconv.Atoi(parts[1])
	}
	if len(parts) >= 3 {
		r.Height, _ = strconv.Atoi(parts[2])
	}
	if len(parts) >= 4 && parts[3] != "" {
		r.Duration, _ = strconv.ParseFloat(parts[3], 64)
	}

	r.Valid = true
	return r
}

// CheckFiles checks integrity of all media files in a folder using
// concurrent goroutines.
func CheckFiles(folder string, maxWorkers int) ([]IntegrityResult, error) {
	allExts := make(map[string]struct{})
	for k, v := range shared.ImageExtensions {
		allExts[k] = v
	}
	for k, v := range shared.VideoExtensions {
		allExts[k] = v
	}

	files, err := shared.FindFilesByExtension(folder, allExts, shared.ExcludedFolders)
	if err != nil {
		return nil, err
	}

	results := make([]IntegrityResult, len(files))
	var wg sync.WaitGroup
	sem := make(chan struct{}, maxWorkers)

	for i, fi := range files {
		wg.Add(1)
		sem <- struct{}{}
		go func(idx int, f shared.FileInfo) {
			defer wg.Done()
			defer func() { <-sem }()

			if shared.IsImage(f.Path) {
				results[idx] = CheckImage(f.Path)
			} else if shared.IsVideo(f.Path) {
				results[idx] = CheckVideo(f.Path)
			} else {
				results[idx] = IntegrityResult{
					Path:  f.Path,
					Type:  "unknown",
					Error: "Unknown file type",
				}
			}
		}(i, fi)
	}

	wg.Wait()
	return results, nil
}
